File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/NetworkStatusHelper.java
package com.brentpanther.bitcoinwidget;

import android.content.Context;
import android.net.ConnectivityManager;
import android.os.Build;
import android.os.PowerManager;

/**
 * Check if anything is restricted preventing the widget from downloading an update.
 */
public class NetworkStatusHelper {

    public static int getRestriction(Context context) {
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PowerManager powerManager = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
            if (powerManager.isPowerSaveMode() && !powerManager.isIgnoringBatteryOptimizations(context.getPackageName())) {
                return 0; //R.string.error_restricted_battery_saver;
            }
        }
        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            int restrictBackgroundStatus = connectivityManager.getRestrictBackgroundStatus();
            if (restrictBackgroundStatus == android.net.ConnectivityManager.RESTRICT_BACKGROUND_STATUS_ENABLED) {
                return R.string.error_restricted_data_saver;
            }
        }
        return 0;
    }
}

File: bitcoin/src/main/java/com/brentpanther/bitcoinwidget/NetworkStatusHelper.kt
package com.brentpanther.bitcoinwidget

import android.content.Context
import android.net.ConnectivityManager
import android.os.Build
import android.os.PowerManager

/**
 * Check if anything is restricted preventing the widget from downloading an update.
 */
object NetworkStatusHelper {

    private fun checkBattery(context: Context): Int {
        val powerManager = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        if (powerManager.isPowerSaveMode && !powerManager.isIgnoringBatteryOptimizations(context.packageName)) {
            return R.string.error_restricted_battery_saver
        }
        return 0
    }

    private fun checkBackgroundData(context: Context): Int {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            if (connectivityManager.restrictBackgroundStatus == ConnectivityManager.RESTRICT_BACKGROUND_STATUS_ENABLED &&
                    connectivityManager.isActiveNetworkMetered) {
                return R.string.error_restricted_data_saver
            }
        }
        return 0
    }

    fun getRestriction(context: Context): Int {
        val checkBattery = checkBattery(context)
        if (checkBattery > 0) return checkBattery
        return checkBackgroundData(context)
    }

}
